File: src/copaw/app/routers/media.py
# -*- coding: utf-8 -*-
from __future__ import annotations

import json

from fastapi import APIRouter, File, Form, HTTPException, Request, UploadFile

from ...media import (
    MediaAnalysisRequest,
    MediaAnalysisResponse,
    MediaAnalysisSummary,
    MediaCapabilityState,
    MediaIngestRequest,
    MediaIngestResponse,
    MediaResolveLinkRequest,
    MediaResolveLinkResponse,
    MediaService,
    MediaSourceSpec,
)

router = APIRouter(prefix="/media", tags=["media"])


def _get_media_service(request: Request) -> MediaService:
    service = getattr(request.app.state, "media_service", None)
    if isinstance(service, MediaService):
        return service
    raise HTTPException(503, detail="Media service is not available")


@router.get("/capabilities", response_model=MediaCapabilityState)
async def get_media_capabilities(request: Request) -> MediaCapabilityState:
    return _get_media_service(request).capabilities()


@router.post("/resolve-link", response_model=MediaResolveLinkResponse)
async def resolve_media_link(
    request: Request,
    payload: MediaResolveLinkRequest,
) -> MediaResolveLinkResponse:
    try:
        return _get_media_service(request).resolve_link(payload)
    except ValueError as exc:
        raise HTTPException(400, detail=str(exc)) from exc


@router.post("/ingest", response_model=MediaIngestResponse)
async def ingest_media(
    request: Request,
    file: UploadFile | None = File(default=None),
    source: str | None = Form(default=None),
) -> MediaIngestResponse:
    service = _get_media_service(request)
    if file is not None:
        try:
            source_payload = json.loads(source) if source else {}
        except json.JSONDecodeError as exc:
            raise HTTPException(400, detail="Invalid media source payload") from exc
        if not isinstance(source_payload, dict):
            raise HTTPException(400, detail="Invalid media source payload")
        payload = MediaIngestRequest(
            source=MediaSourceSpec.model_validate(
                {"source_kind": "upload", **source_payload},
            )
        )
        try:
            return service.ingest(
                payload,
                file_bytes=await file.read(),
                filename=file.filename,
                mime_type=file.content_type,
            )
        except ValueError as exc:
            raise HTTPException(400, detail=str(exc)) from exc

    try:
        raw_payload = await request.json()
    except Exception as exc:
        raise HTTPException(400, detail="Media ingest requires JSON or multipart form data") from exc
    try:
        payload = MediaIngestRequest.model_validate(raw_payload)
        return service.ingest(payload)
    except ValueError as exc:
        raise HTTPException(400, detail=str(exc)) from exc


@router.post("/analyses", response_model=MediaAnalysisResponse)
async def analyze_media(
    request: Request,
    payload: MediaAnalysisRequest,
) -> MediaAnalysisResponse:
    try:
        return await _get_media_service(request).analyze(payload)
    except ValueError as exc:
        raise HTTPException(400, detail=str(exc)) from exc


@router.get("/analyses", response_model=list[MediaAnalysisSummary])
async def list_media_analyses(
    request: Request,
    industry_instance_id: str | None = None,
    thread_id: str | None = None,
    entry_point: str | None = None,
    status: str | None = None,
    limit: int = 50,
) -> list[MediaAnalysisSummary]:
    return _get_media_service(request).list_analyses(
        industry_instance_id=industry_instance_id,
        thread_id=thread_id,
        entry_point=entry_point,
        status=status,
        limit=limit,
    )


@router.get("/analyses/{analysis_id}", response_model=MediaAnalysisSummary)
async def get_media_analysis(
    request: Request,
    analysis_id: str,
) -> MediaAnalysisSummary:
    summary = _get_media_service(request).get_analysis(analysis_id)
    if summary is None:
        raise HTTPException(404, detail=f"Media analysis '{analysis_id}' not found")
    return summary

File: console/src/api/modules/media.ts
import { authenticatedFetch, request } from "../request";

export type MediaSourceKind = "link" | "upload" | "existing-artifact";
export type MediaType = "unknown" | "article" | "video" | "audio" | "document";
export type AnalysisMode = "standard" | "video-lite" | "video-deep";
export type MediaEntryPoint =
  | "industry-preview"
  | "industry-bootstrap"
  | "chat"
  | "runtime-center";
export type MediaPurpose =
  | "draft-enrichment"
  | "chat-answer"
  | "learn-and-writeback"
  | "reference-only";

export interface MediaCapabilityState {
  video_deep_available: boolean;
  native_video_enabled: boolean;
  native_audio_enabled: boolean;
  local_asr_enabled: boolean;
  supported_video_modes: AnalysisMode[];
}

export interface MediaSourceSpec {
  source_id?: string;
  source_kind: MediaSourceKind;
  media_type?: MediaType;
  declared_media_type?: MediaType | null;
  detected_media_type?: MediaType | null;
  analysis_mode?: AnalysisMode | null;
  title?: string | null;
  url?: string | null;
  filename?: string | null;
  mime_type?: string | null;
  size_bytes?: number | null;
  artifact_id?: string | null;
  storage_uri?: string | null;
  upload_base64?: string | null;
  entry_point?: MediaEntryPoint;
  purpose?: MediaPurpose;
  metadata?: Record<string, unknown>;
}

export interface MediaResolveLinkPayload {
  url: string;
  entry_point?: MediaEntryPoint;
  purpose?: MediaPurpose;
}

export interface MediaResolveLinkResponse {
  url: string;
  normalized_url: string;
  detected_media_type: MediaType;
  mime_type?: string | null;
  title?: string | null;
  filename?: string | null;
  size_bytes?: number | null;
  analysis_mode_options: AnalysisMode[];
  resolved_source: MediaSourceSpec;
  warnings: string[];
  capabilities: MediaCapabilityState;
}

export interface MediaIngestResponse {
  source: MediaSourceSpec;
  detected_media_type: MediaType;
  analysis_mode_options: AnalysisMode[];
  asset_artifact_ids: string[];
  evidence_ids: string[];
  warnings: string[];
  capabilities: MediaCapabilityState;
}

export interface MediaAnalysisSummary {
  analysis_id: string;
  industry_instance_id?: string | null;
  thread_id?: string | null;
  entry_point: string;
  purpose: string;
  source_kind: string;
  source_ref?: string | null;
  detected_media_type: MediaType | string;
  analysis_mode: AnalysisMode | string;
  status: string;
  title: string;
  url?: string | null;
  filename?: string | null;
  mime_type?: string | null;
  size_bytes?: number | null;
  summary: string;
  key_points: string[];
  entities: string[];
  claims: string[];
  recommended_actions: string[];
  warnings: string[];
  asset_artifact_ids: string[];
  derived_artifact_ids: string[];
  transcript_artifact_id?: string | null;
  knowledge_document_ids: string[];
  evidence_ids: string[];
  strategy_writeback_status?: string | null;
  backlog_writeback_status?: string | null;
  error_message?: string | null;
  metadata: Record<string, unknown>;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface MediaAnalysisPayload {
  sources: MediaSourceSpec[];
  industry_instance_id?: string | null;
  thread_id?: string | null;
  entry_point?: MediaEntryPoint;
  purpose?: MediaPurpose;
  writeback?: boolean;
}

export interface MediaAnalysisResponse {
  analyses: MediaAnalysisSummary[];
  warnings: string[];
  capabilities: MediaCapabilityState;
}

async function uploadIngestedMedia(
  source: MediaSourceSpec,
  file: File,
): Promise<MediaIngestResponse> {
  const body = new FormData();
  body.set("source", JSON.stringify(source));
  body.set("file", file);
  const response = await authenticatedFetch("/media/ingest", {
    method: "POST",
    body,
  });
  return (await response.json()) as MediaIngestResponse;
}

export const mediaApi = {
  getMediaCapabilities(): Promise<MediaCapabilityState> {
    return request<MediaCapabilityState>("/media/capabilities");
  },

  resolveMediaLink(
    payload: MediaResolveLinkPayload,
  ): Promise<MediaResolveLinkResponse> {
    return request<MediaResolveLinkResponse>("/media/resolve-link", {
      method: "POST",
      body: JSON.stringify(payload),
    });
  },

  ingestMedia(
    source: MediaSourceSpec,
    file?: File | null,
  ): Promise<MediaIngestResponse> {
    if (file) {
      return uploadIngestedMedia(source, file);
    }
    return request<MediaIngestResponse>("/media/ingest", {
      method: "POST",
      body: JSON.stringify({ source }),
    });
  },

  analyzeMedia(payload: MediaAnalysisPayload): Promise<MediaAnalysisResponse> {
    return request<MediaAnalysisResponse>("/media/analyses", {
      method: "POST",
      body: JSON.stringify(payload),
    });
  },

  listMediaAnalyses(params?: {
    industry_instance_id?: string;
    thread_id?: string;
    entry_point?: string;
    status?: string;
    limit?: number;
  }): Promise<MediaAnalysisSummary[]> {
    const query = new URLSearchParams();
    if (params?.industry_instance_id) {
      query.set("industry_instance_id", params.industry_instance_id);
    }
    if (params?.thread_id) {
      query.set("thread_id", params.thread_id);
    }
    if (params?.entry_point) {
      query.set("entry_point", params.entry_point);
    }
    if (params?.status) {
      query.set("status", params.status);
    }
    if (typeof params?.limit === "number") {
      query.set("limit", String(params.limit));
    }
    return request<MediaAnalysisSummary[]>(
      `/media/analyses${query.toString() ? `?${query.toString()}` : ""}`,
    );
  },

  getMediaAnalysis(analysisId: string): Promise<MediaAnalysisSummary> {
    return request<MediaAnalysisSummary>(`/media/analyses/${analysisId}`);
  },
};
